### PR TITLE
feat(storage): add Custom Placement Config Dual Region Support 

### LINKED ISSUE #6116: storage: Add Support for Updated Dual-Region Bucket API
[Internal]
- For details and coordination: [go/gcs-dpe-dual-region-milestones-api-refactor](http://go/gcs-dpe-dual-region-milestones-api-refactor)
- Implementation Example: https://github.com/googleapis/nodejs-storage/pull/1977

### PR DESCRIPTION
**Pull request by cojenco**
Fixes #6116 

Note: Tests pending new config to be rolled out by July 6th, 2022

### CODE REVIEW

**Review by danielbankhead — APPROVED** [review]

File: storage/bucket.go [hunk 1]
```diff
@@ -533,6 +534,9 @@ type BucketAttrs struct {
 	// See https://cloud.google.com/storage/docs/managing-turbo-replication for
 	// more information.
 	RPO RPO
+
+	// The bucket's custom placement configuration for Custom Dual Regions.
```
Comment (tritone) [inline storage/bucket.go:538]: I would add the word "location" somewhere in this to make it clear that it's tied to the location. Maybe it should go next to `Location` as well.

File: storage/bucket.go [hunk 2]
```diff
@@ -782,6 +786,15 @@ type BucketWebsite struct {
 	NotFoundPage string
 }
 
+// BucketCustomPlacementConfig holds the bucket's custom placement
+// configuration for Custom Dual Regions. See
+// https://cloud.google.com/storage/docs/locations for more information.
```
Comment (tritone) [inline storage/bucket.go:791]: Is there a subsection we can link to in the future that's specifically about CDR? @danielbankhead

File: storage/bucket.go [hunk 3]
```diff
@@ -1904,6 +1921,38 @@ func toRPOFromProto(b *storagepb.Bucket) RPO {
 	}
 }
 
+func toBucketCustomPlacementConfig(c *raw.BucketCustomPlacementConfig) *BucketCustomPlacementConfig {
```
Comment (tritone) [inline storage/bucket.go:1924]: For parallelism maybe let's call them `customPlacementFromRaw` and `customPlacementFromProto`?

File: storage/bucket.go [hunk 4]
```diff
@@ -782,6 +786,15 @@ type BucketWebsite struct {
 	NotFoundPage string
 }
 
+// BucketCustomPlacementConfig holds the bucket's custom placement
+// configuration for Custom Dual Regions. See
+// https://cloud.google.com/storage/docs/locations for more information.
+type BucketCustomPlacementConfig struct {
```
Comment (tritone) [inline storage/bucket.go:792]: Perhaps just `CustomPlacementConfig` for brevity (and I think that's what it's called in the API itself, the `Bucket` is just from the generated code).

**Review by tritone — COMMENTED** [review]
Few naming things, overall looks great!

File: storage/bucket.go (hunk 4 above)
Comment (cojenco) [reply storage/bucket.go]: sounds good, updated!

File: storage/bucket.go (hunk 2 above)
Comment (danielbankhead) [reply storage/bucket.go:792]: Yep, here's the subsection: https://cloud.google.com/storage/docs/locations#location-dr

Comment (tritone) [reply storage/bucket.go:792]: Cool -- @cojenco if you can update to use that link that would be great.

Comment (cojenco) [reply storage/bucket.go]: awesome, thanks folks!

**Review by tritone — APPROVED** [review]

**Pull request merged**